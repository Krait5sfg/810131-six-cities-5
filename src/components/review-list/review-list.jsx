import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {ReviewPropTypes} from '../../utils/property-type';
import Review from '../review/review';
import {connect} from 'react-redux';
import {getActiveOfferComments} from '../../store/api-actions';

const LIMIT_COMMENT = 10;

class ReviewList extends PureComponent {

  componentDidMount() {
    const {updateActiveOfferComments, idActiveOffer} = this.props;
    updateActiveOfferComments(idActiveOffer);
  }

  render() {
    const {activeOfferComments} = this.props;

    // сортировка комментариев чтобы сначала рендерились по свежей дате
    const sortedCommentsByDate = activeOfferComments.slice().sort((first, second) => first.date > second.date ? -1 : 1).slice(0, LIMIT_COMMENT);

    const commentCount = activeOfferComments.length;
    const commentElements = sortedCommentsByDate.map((comment, id) => <Review comment={comment} key={id} />);

    return (
      <React.Fragment>
        <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{commentCount}</span></h2>
        <ul className="reviews__list">
          {commentElements}
        </ul>
      </React.Fragment>
    );
  }
}

ReviewList.propTypes = {
  activeOfferComments: PropTypes.arrayOf(ReviewPropTypes).isRequired,
  updateActiveOfferComments: PropTypes.func.isRequired,
  idActiveOffer: PropTypes.number.isRequired
};

const mapStateToProps = (({DATA}) => ({
  activeOfferComments: DATA.activeOfferComments,
}));

const mapDispatchToProps = ((dispatch) => ({
  updateActiveOfferComments(idActiveOffer) {
    dispatch(getActiveOfferComments(idActiveOffer));
  }
}));

export {ReviewList};
export default connect(mapStateToProps, mapDispatchToProps)(ReviewList);
