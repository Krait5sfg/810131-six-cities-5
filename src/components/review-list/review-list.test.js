import React from 'react';
import renderer from 'react-test-renderer';
import {ReviewList} from './review-list';


const comments = [
  {
    avatar: ``,
    date: new Date(`10-08-2020`),
    // id: 1,
    isPro: true,
    // commentText: ``,
    name: ``,
    rating: 0,
  },
  {
    avatar: ``,
    date: new Date(`11-08-2020`),
    // id: 1,
    isPro: true,
    // commentText: ``,
    name: ``,
    rating: 0,
  },
];

jest.mock(`../review/review`, () => `Review`);

it(`ReviewList render correctly`, () => {
  const tree = renderer.create(<ReviewList
    activeOfferComments={comments}
    updateActiveOfferComments={() => {}}
    idActiveOffer={1}
  />).toJSON();
  expect(tree).toMatchSnapshot();
});
