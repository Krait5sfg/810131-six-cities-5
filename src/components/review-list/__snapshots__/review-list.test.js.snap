// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewList render correctly 1`] = `
Array [
  <h2
    className="reviews__title"
  >
    Reviews Â· 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>,
  <ul
    className="reviews__list"
  >
    <Review
      comment={
        Object {
          "avatar": "",
          "date": 2020-11-07T21:00:00.000Z,
          "isPro": true,
          "name": "",
          "rating": 0,
        }
      }
    />
    <Review
      comment={
        Object {
          "avatar": "",
          "date": 2020-10-07T21:00:00.000Z,
          "isPro": true,
          "name": "",
          "rating": 0,
        }
      }
    />
  </ul>,
]
`;
